File: android/src/main/kotlin/com/worldcredit/badge/ui/InlineBadge.kt
package com.worldcredit.badge.ui

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import com.worldcredit.badge.BadgeData
import com.worldcredit.badge.BadgeSize
import com.worldcredit.badge.BadgeState
import com.worldcredit.badge.BadgeTier
import com.worldcredit.badge.WorldCreditBadge
import com.worldcredit.badge.getScoreDisplay
import com.worldcredit.badge.rememberBadgeState
import com.worldcredit.badge.theme.BadgeColors
import com.worldcredit.badge.theme.BadgeDimensions
import com.worldcredit.badge.theme.BadgeTheme
import com.worldcredit.badge.theme.BadgeTypography

/**
 * InlineBadge - Tiny pill badge that sits inline next to text
 * Format: [WC logo] 52 · Gold
 * 
 * @param handle User handle to fetch and display badge for
 * @param theme Badge theme (Light, Dark, Auto)
 * @param size Badge size scaling
 * @param modifier Compose modifier
 */
@Composable
fun InlineBadge(
    handle: String,
    theme: BadgeTheme = BadgeTheme.AUTO,
    size: BadgeSize = BadgeSize.MEDIUM,
    modifier: Modifier = Modifier
) {
    val badgeState = rememberBadgeState(handle)
    
    InlineBadge(
        state = badgeState,
        theme = theme,
        size = size,
        modifier = modifier
    )
}

/**
 * InlineBadge with direct state management
 * 
 * @param state Current badge state
 * @param theme Badge theme (Light, Dark, Auto)
 * @param size Badge size scaling
 * @param modifier Compose modifier
 */
@Composable
fun InlineBadge(
    state: BadgeState,
    theme: BadgeTheme = BadgeTheme.AUTO,
    size: BadgeSize = BadgeSize.MEDIUM,
    modifier: Modifier = Modifier
) {
    val colors = BadgeColors.forTheme(theme)
    val context = LocalContext.current
    val sizeScale = size.scale
    
    Box(
        modifier = modifier
            .height(BadgeDimensions.inlineHeight * sizeScale)
            .clip(RoundedCornerShape(BadgeDimensions.cornerRadius * sizeScale))
            .background(colors.surface)
            .padding(horizontal = BadgeDimensions.inlinePadding * sizeScale),
        contentAlignment = Alignment.Center
    ) {
        when (state) {
            is BadgeState.Loading -> {
                InlineBadgeLoading(colors = colors, sizeScale = sizeScale)
            }
            is BadgeState.Success -> {
                InlineBadgeContent(
                    badgeData = state.data,
                    colors = colors,
                    sizeScale = sizeScale,
                    onClick = { WorldCreditBadge.openProfile(context, state.data) }
                )
            }
            is BadgeState.Error -> {
                InlineBadgeError(colors = colors, sizeScale = sizeScale)
            }
        }
    }
}

@Composable
private fun InlineBadgeContent(
    badgeData: BadgeData,
    colors: BadgeColors.BadgeColorScheme,
    sizeScale: Float,
    onClick: () -> Unit
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(BadgeDimensions.inlineSpacing * sizeScale),
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier.clickable { onClick() }
    ) {
        // World Credit Logo
        AsyncImage(
            model = WorldCreditBadge.getLogoUrl(),
            contentDescription = "World Credit Logo",
            modifier = Modifier
                .size(BadgeDimensions.logoSize * sizeScale)
                .alpha(if (badgeData.isUnverified) 0.5f else 1f)
        )
        
        if (badgeData.isUnverified) {
            // "Not Verified" text in muted color
            Text(
                text = "Not Verified",
                fontSize = BadgeTypography.scoreTextSize * sizeScale,
                fontWeight = FontWeight.Medium,
                color = colors.onSurfaceVariant
            )
        } else {
            // Score and tier text
            Text(
                text = badgeData.getScoreDisplay(),
                fontSize = BadgeTypography.scoreTextSize * sizeScale,
                fontWeight = FontWeight.Medium,
                color = colors.onSurface
            )
            
            // Tier color indicator
            Box(
                modifier = Modifier
                    .size(6.dp * sizeScale)
                    .background(
                        color = BadgeColors.getTierColor(badgeData.tierEnum),
                        shape = RoundedCornerShape(50)
                    )
            )
        }
    }
}

@Composable
private fun InlineBadgeLoading(
    colors: BadgeColors.BadgeColorScheme,
    sizeScale: Float
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(BadgeDimensions.inlineSpacing * sizeScale),
        verticalAlignment = Alignment.CenterVertically
    ) {
        // Placeholder for logo
        Box(
            modifier = Modifier
                .size(BadgeDimensions.logoSize * sizeScale)
                .background(
                    color = colors.surfaceVariant,
                    shape = RoundedCornerShape(4.dp * sizeScale)
                )
        )
        
        // Loading indicator
        CircularProgressIndicator(
            modifier = Modifier.size(12.dp * sizeScale),
            strokeWidth = 1.5.dp * sizeScale,
            color = colors.onSurfaceVariant
        )
        
        // Animated placeholder text
        val infiniteTransition = rememberInfiniteTransition(label = "loading")
        val alpha by infiniteTransition.animateFloat(
            initialValue = 0.3f,
            targetValue = 0.8f,
            animationSpec = infiniteRepeatable(
                animation = tween(1000, easing = LinearEasing),
                repeatMode = RepeatMode.Reverse
            ),
            label = "alpha"
        )
        
        Text(
            text = "Loading...",
            fontSize = BadgeTypography.scoreTextSize * sizeScale,
            color = colors.onSurfaceVariant.copy(alpha = alpha)
        )
    }
}

@Composable
private fun InlineBadgeError(
    colors: BadgeColors.BadgeColorScheme,
    sizeScale: Float
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(BadgeDimensions.inlineSpacing * sizeScale),
        verticalAlignment = Alignment.CenterVertically
    ) {
        // Error icon placeholder
        Box(
            modifier = Modifier
                .size(BadgeDimensions.logoSize * sizeScale)
                .background(
                    color = Color(0xFFFF6B6B),
                    shape = RoundedCornerShape(4.dp * sizeScale)
                )
        )
        
        Text(
            text = "Error",
            fontSize = BadgeTypography.scoreTextSize * sizeScale,
            color = Color(0xFFFF6B6B)
        )
    }
}

@Preview(name = "Inline Badge Success")
@Composable
private fun InlineBadgeSuccessPreview() {
    InlineBadge(
        state = BadgeState.Success(
            BadgeData(
                ok = true,
                handle = "janedoe",
                displayName = "Jane Doe",
                worldScore = 75,
                tier = "Gold",
                tierColor = "#FFD700",
                photoUrl = null,
                linkedNetworks = listOf("twitter", "linkedin"),
                profileUrl = "https://worldcredit.com/janedoe",
                categories = listOf("tech", "finance")
            )
        ),
        theme = BadgeTheme.LIGHT
    )
}

@Preview(name = "Inline Badge Loading")
@Composable
private fun InlineBadgeLoadingPreview() {
    InlineBadge(
        state = BadgeState.Loading,
        theme = BadgeTheme.LIGHT
    )
}

@Preview(name = "Inline Badge Error")
@Composable
private fun InlineBadgeErrorPreview() {
    InlineBadge(
        state = BadgeState.Error("Network error"),
        theme = BadgeTheme.LIGHT
    )
}

@Preview(name = "Inline Badge Dark")
@Composable
private fun InlineBadgeDarkPreview() {
    InlineBadge(
        state = BadgeState.Success(
            BadgeData(
                ok = true,
                handle = "sarahk",
                displayName = "Sarah K.",
                worldScore = 89,
                tier = "Platinum",
                tierColor = "#00FFC8",
                photoUrl = null,
                linkedNetworks = listOf("github"),
                profileUrl = "https://worldcredit.com/sarahk",
                categories = listOf("opensource")
            )
        ),
        theme = BadgeTheme.DARK
    )
}

File: ios/Sources/WorldCreditBadge/Views/InlineBadge.swift
import SwiftUI

/// Tiny pill badge that sits inline next to text: [WC logo] 52 · Gold
public struct InlineBadge: View {
    private let handle: String
    private let email: String?
    private let theme: BadgeTheme
    private let size: BadgeSize
    
    @StateObject private var dataModel: BadgeDataModel
    
    /// Initialize with handle and optional customization
    /// - Parameters:
    ///   - handle: User handle to display badge for
    ///   - theme: Visual theme (default: automatic)
    ///   - size: Badge size (default: small)
    public init(
        handle: String = "",
        email: String? = nil,
        theme: BadgeTheme = .automatic,
        size: BadgeSize = .sm
    ) {
        self.handle = handle
        self.email = email
        self.theme = theme
        self.size = size
        self._dataModel = StateObject(wrappedValue: BadgeDataModel(handle: handle, email: email))
    }
    
    /// Initialize with pre-fetched badge data
    /// - Parameters:
    ///   - badgeData: Pre-fetched badge data
    ///   - theme: Visual theme (default: automatic)
    ///   - size: Badge size (default: small)
    public init(
        badgeData: BadgeData,
        theme: BadgeTheme = .automatic,
        size: BadgeSize = .sm
    ) {
        self.handle = badgeData.handle
        self.theme = theme
        self.size = size
        self._dataModel = StateObject(wrappedValue: BadgeDataModel(handle: badgeData.handle))
        
        // Set the pre-fetched data
        DispatchQueue.main.async {
            self.dataModel.badgeData = badgeData
        }
    }
    
    public var body: some View {
        Group {
            if let badgeData = dataModel.badgeData {
                BadgeContent(badgeData: badgeData, theme: theme, size: size)
            } else if dataModel.isLoading {
                LoadingContent(theme: theme, size: size)
            } else if dataModel.error != nil {
                ErrorContent(theme: theme, size: size)
            } else {
                EmptyView()
            }
        }
        .onAppear {
            if dataModel.badgeData == nil && !dataModel.isLoading {
                dataModel.loadData()
            }
        }
    }
}

private struct BadgeContent: View {
    let badgeData: BadgeData
    let theme: BadgeTheme
    let size: BadgeSize
    
    var body: some View {
        HStack(spacing: 4) {
            // World Credit logo
            WorldCreditLogo(size: size.iconSize)
                .opacity(badgeData.isUnverified ? 0.5 : 1.0)
            
            // Score and tier text (or "Not Verified")
            if badgeData.isUnverified {
                Text("Not Verified")
                    .font(.system(size: size.rawValue * 0.6, weight: .medium))
                    .foregroundColor(theme.secondaryTextColor)
                    .lineLimit(1)
            } else {
                Text(badgeData.scoreDisplayText)
                    .font(.system(size: size.rawValue * 0.6, weight: .medium))
                    .foregroundColor(theme.textColor)
                    .lineLimit(1)
            }
        }
        .padding(.horizontal, 6)
        .padding(.vertical, 2)
        .background(
            Capsule()
                .fill(theme.backgroundColor)
                .overlay(
                    Capsule()
                        .stroke(
                            badgeData.isUnverified
                                ? Color.gray.opacity(0.3)
                                : badgeData.tierType.color.opacity(0.3),
                            lineWidth: 1
                        )
                )
        )
        .profileTappable(badgeData)
    }
}

private struct LoadingContent: View {
    let theme: BadgeTheme
    let size: BadgeSize
    
    var body: some View {
        HStack(spacing: 4) {
            ProgressView()
                .scaleEffect(0.7)
                .frame(width: size.iconSize, height: size.iconSize)
            
            Text("•••")
                .font(.system(size: size.rawValue * 0.6, weight: .medium))
                .foregroundColor(theme.secondaryTextColor)
                .lineLimit(1)
        }
        .padding(.horizontal, 6)
        .padding(.vertical, 2)
        .background(
            Capsule()
                .fill(theme.backgroundColor)
                .overlay(
                    Capsule()
                        .stroke(theme.borderColor, lineWidth: 1)
                )
        )
    }
}

private struct ErrorContent: View {
    let theme: BadgeTheme
    let size: BadgeSize
    
    var body: some View {
        HStack(spacing: 4) {
            Image(systemName: "exclamationmark.triangle.fill")
                .foregroundColor(.orange)
                .font(.system(size: size.iconSize * 0.8))
            
            Text("Error")
                .font(.system(size: size.rawValue * 0.6, weight: .medium))
                .foregroundColor(theme.secondaryTextColor)
                .lineLimit(1)
        }
        .padding(.horizontal, 6)
        .padding(.vertical, 2)
        .background(
            Capsule()
                .fill(theme.backgroundColor)
                .overlay(
                    Capsule()
                        .stroke(theme.borderColor, lineWidth: 1)
                )
        )
    }
}

// MARK: - Previews

struct InlineBadge_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: 16) {
            // Sample usage in text context
            HStack {
                Text("@janedoe")
                    .font(.headline)
                InlineBadge(
                    badgeData: BadgeData(
                        ok: true,
                        handle: "janedoe",
                        displayName: "Jane Doe",
                        worldScore: 52,
                        tier: "Gold",
                        tierColor: "#FFD700",
                        photoUrl: nil,
                        linkedNetworks: nil,
                        profileUrl: "https://world-credit.com/janedoe",
                        categories: nil
                    )
                )
                Spacer()
            }
            
            // Different sizes
            VStack(alignment: .leading, spacing: 8) {
                ForEach(BadgeSize.allCases, id: \.rawValue) { size in
                    HStack {
                        Text("Size \(size.rawValue, specifier: "%.0f"):")
                            .font(.caption)
                            .foregroundColor(.secondary)
                        InlineBadge(
                            badgeData: BadgeData(
                                ok: true,
                                handle: "sarah_k",
                                displayName: "Sarah K.",
                                worldScore: 85,
                                tier: "Platinum",
                                tierColor: "#00FFC8",
                                photoUrl: nil,
                                linkedNetworks: nil,
                                profileUrl: "https://world-credit.com/sarah_k",
                                categories: nil
                            ),
                            size: size
                        )
                        Spacer()
                    }
                }
            }
            
            // Different themes
            VStack(alignment: .leading, spacing: 8) {
                HStack {
                    Text("Light:")
                        .font(.caption)
                        .foregroundColor(.secondary)
                    InlineBadge(
                        badgeData: BadgeData(
                            ok: true,
                            handle: "testuser",
                            displayName: "Test User",
                            worldScore: 25,
                            tier: "Silver",
                            tierColor: "#C0C0C0",
                            photoUrl: nil,
                            linkedNetworks: nil,
                            profileUrl: "https://world-credit.com/testuser",
                            categories: nil
                        ),
                        theme: .light
                    )
                    Spacer()
                }
                
                HStack {
                    Text("Dark:")
                        .font(.caption)
                        .foregroundColor(.secondary)
                    InlineBadge(
                        badgeData: BadgeData(
                            ok: true,
                            handle: "testuser",
                            displayName: "Test User",
                            worldScore: 25,
                            tier: "Silver",
                            tierColor: "#C0C0C0",
                            photoUrl: nil,
                            linkedNetworks: nil,
                            profileUrl: "https://world-credit.com/testuser",
                            categories: nil
                        ),
                        theme: .dark
                    )
                    Spacer()
                }
            }
            
            Spacer()
        }
        .padding()
    }
}